api/v1alpha1/stargate_types.go

```diff
@@ -0,0 +1,143 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StargateSpec defines the desired state of Stargate
+type StargateSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// DatacenterName is the name of a CassandraDatacenter resource where Stargate should be
+	// deployed.
+	// +kubebuilder:validation:Required
+	DatacenterName string `json:"datacenterName"`
+
+	// Size is the number of Stargate instances to deploy. This value may be scaled independently of
+	// Cassandra cluster nodes. Each instance handles API and coordination tasks for inbound
+	// queries.
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:default=1
+	Size int32 `json:"size"`
+
+	// ServerVersion is the version of Stargate to deploy.
+	// +kubebuilder:validation:Pattern=\d+\.\d+\.\d+
+	// +kubebuilder:default="1.0.29"
+	// FIXME can we use type container.Version or version.Version here?
+	ServerVersion string `json:"serverVersion"`
+
+	// ContainerImage is the image characteristics to use for Stargate containers. Leave nil to use
+	// a default image.
+	// +optional
+	ContainerImage *ContainerImage `json:"containerImage,omitempty"`
+
+	// InitContainerImage is the image characteristics to use for Stargate init containers. Init
+	// containers are used to wait for Cassandra. Leave nil to use a default image.
+	// +optional
+	InitContainerImage *ContainerImage `json:"initContainerImage,omitempty"`
+
+	// ServiceAccount is the service account name to use for Stargate pods.
+	// +kubebuilder:default="default"
+	ServiceAccount string `json:"serviceAccount"`
+
+	// Resources is the Kubernetes resource requests and limits to apply, per Stargate pod. Leave
+	// nil to use defaults.
+	// +optional
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+
+	// HeapSize sets the JVM heap size to use for Stargate. If no Resources are specified, this
+	// value will also be used to set a default memory request and limit for the Stargate pods:
+	// these will be set to HeapSize x2 and x4, respectively.
+	// +kubebuilder:default="256Mi"
+	HeapSize resource.Quantity `json:"heapSize"`
+
+	// LivenessProbe sets the Stargate liveness probe. Leave nil to use defaults.
+	// +optional
+	LivenessProbe *corev1.Probe `json:"livenessProbe,omitempty"`
+
+	// ReadinessProbe sets the Stargate readiness probe. Leave nil to use defaults.
+	// +optional
+	ReadinessProbe *corev1.Probe `json:"readinessProbe,omitempty"`
+
+	// Affinity is the affinity to apply to the Stargate pods. See
+	// https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity
+	// for background.
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty"`
```

Comment:
Based on prior discussions with @adejanovski my understanding is that we want to deploy Stargate nodes in each rack. Affinity for Cassandra pods is configured implicitly via racks. I think we will want piggy back off of that. By default, I think we should look at the rack configuration of the CassandraDatacenter and configure affinity such that we will get Stargate pods distributed evenly among racks (as best we can). This goes beyond the scope of this PR, but I wanted to at least point it out here.